1Password/shell-plugins
# AWS Vault importer for the AWS shell plugin

## Summary

[AWS Vault](https://github.com/99designs/aws-vault) is a popular tool among AWS users used to store/provision AWS credentials in a vaulting backend such as macOS Keychain. Chances are if we want to convince an AWS user to also use the AWS shell plugin then they would need to migrate their credentials from AWS Vault. Therefore, we need to create an importer for the AWS access keys, that would fetch those from the user's vaulting backend and put them in 1Password. Luckily, AWS Vault is also open source, so we can leverage their own functionality for retrieving credentials.

## Thought Process

When a profile is added through `aws-vault`, a section is added to the AWS config file usually located at `~/.aws/config`. The AWS Vault importer iterates through the config file, searches the user's vaulting backend for matching credentials (taking advantage of functionality defined in the AWS Vault repo), then imports them to 1Password.

~~Since AWS Vault relies on the AWS config file to function, designing the importer as a file importer seemed intuitive.~~

The AWS region name for an AWS profile can be specified in `~/.aws/config`, for specific profiles and/or for the `default` profile. When an AWS Vault profile is imported, the region that's used for the item is determined as follows:

| AWS Vault profile region specified? | `default` profile region specified? | Region used |
|-------------------------------------|-------------------------------------|-------------|
| Yes                                 | Yes                                 | AWS Vault profile   |
| Yes                                 | No                                  | AWS Vault profile   |
| No                                  | Yes                                 | `default`   |
| No                                  | No                                  | None        |

## How To Test

1. Install `aws-vault` following the instructions outlined [here](https://github.com/99designs/aws-vault)
2. Add an AWS profile to your keychain through `aws-vault`, i.e. `aws-vault add <profile>`
3. Run `op plugin init aws` > `Import into 1Password...` > enter your keychain password(s) when prompted. You should see options to import from `<keychain> (<profile>)`. If there are multiple AWS Vault profiles, each one should be listed
4. Select a profile to import and complete the import flow. An `AWS Access Key (<profile>)` item should be created in 1Password
5. Run an `aws` command that requires authentication, e.g. `aws s3 ls`, and use the newly created item to authenticate. The command should run successfully
6. Test that the logic for determining the item's region, outlined in "Thought Process", is followed

![image](https://user-images.githubusercontent.com/53918631/229199831-1599b6a4-3e61-4d7b-bd78-fa5c05f26b88.png)
